src/object/traverse.ts

```diff
@@ -0,0 +1,199 @@
+import { isArray, isFunction, isIterable, isPlainObject, last } from 'radashi'
+
+/**
+ * Iterate an object's properties and those of any nested objects.
+ * “Non-array iterables” (e.g. Map and Set instances) are only
+ * traversed when they are the root object.
+ *
+ * - By default, only plain objects and arrays are traversed.
+ *
+ * - The traversal is performed in a depth-first manner.
+ *
+ * - **Non-Enumerable Properties**  \
+ *   Only enumerable properties are traversed by default. The
+ *   `ownKeys` callback can be customized to traverse non-enumerable
+ *   properties (try passing `Reflect.ownKeys`).
+ *
+ * - **Early Return**  \
+ *   The `visitor` callback can return `false` to stop the traversal.
+ *
+ * - **Skipping**  \
+ *   To skip traversal of an object, call `context.skip(obj)`. If the
+ *   current value is an object, you may call `context.skip()` without
+ *   an argument to skip it.
+ *
+ * - **Nested Iterables & Class Instances**  \
+ *   To traverse a nested iterable or class instance, you can call
+ *   `traverse` inside your `visitor` callback recursively, but make
+ *   sure to pass the `context` object to it.
+ */
+export function traverse(
+  root: object,
+  visitor: TraverseVisitor,
+  outerContext?: TraverseContext | null,
```

Comment:
Do we want to make `outerContext` a public API? Do you have any nice use case examples for that?

I think it's better to create `internalTaverse` which is not exported and export `traverse` as a root call of `internalTraverse`.  It helps us to keep API simple. And we can always add `outerContext` argument in the future if we have nice use case for that.

Reply:
It's intentional. You can call `traverse` from within another `traverse` visitor if you need to traverse a class instance or a non-array iterable. By passing the context along, you maintain the `context.parents` and `context.path` properties.